staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1196,15 +1241,13 @@ func (c *cacheWatcher) convertToWatchEvent(event *watchCacheEvent) *watch.Event
 
 	switch {
 	case curObjPasses && !oldObjPasses:
-		return &watch.Event{Type: watch.Added, Object: event.Object.DeepCopyObject()}
+		return &watch.Event{Type: watch.Added, Object: getEventObject(event.Object)}
 	case curObjPasses && oldObjPasses:
-		return &watch.Event{Type: watch.Modified, Object: event.Object.DeepCopyObject()}
+		return &watch.Event{Type: watch.Modified, Object: getEventObject(event.Object)}
 	case !curObjPasses && oldObjPasses:
 		// return a delete event with the previous object content, but with the event's resource version
-		oldObj := event.PrevObject.DeepCopyObject()
-		if err := c.versioner.UpdateObject(oldObj, event.ResourceVersion); err != nil {
```

Comment:
hmm... this is fragile and potentially problematic (the same original object+resourceVersion can be encoded with different resourceVersions)... definitely need a test that exercises an object at a given resourceVersion getting encoded into delete events at different resourceVersions. For example:
* start three watchers
  * unfiltered
  * `labelSelector=foo=true`
  * `labelSelector=bar=true`
* add object with labels `foo=true`, `bar=true` at rv=1
  * watch events with rv=1 get sent to all watchers
* update object to remove `foo=true` label at rv=2
  * update event with rv=2 and `bar=true` content gets sent to unfiltered and `bar=true` watchers
  * delete event with rv=2 and `foo=true,bar=true` content gets sent to `foo=true` watcher
* update object to remove `bar=true` label at rv=3
  * update event with rv=3 and no label content gets sent to unfiltered watcher
  * delete event with rv=3 and `bar=true` content gets sent to `bar=true` watcher
* delete object at rv=4
  * delete event with rv=4 and no label content gets sent to unfiltered watcher

Reply:
Yeah - this is also the reason why I think the deep-copy is needed.
Will work on adding this test now.

Reply:
done